notifier/amqp/deliverer.go

```diff
@@ -0,0 +1,77 @@
+package amqp
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"path"
+
+	"github.com/google/uuid"
+	clairerror "github.com/quay/clair/v4/clair-error"
+	"github.com/quay/clair/v4/notifier"
+	samqp "github.com/streadway/amqp"
+)
+
+// Deliverer is an AMQP deliverer which publishes a notifier.Callback to the
+// the broker.
+//
+// It's an error to configure this deliverer with an Exchange that does not exist.
+// Administrators should configure the Exchange, Queue, and Bindings before starting
+// this deliverer.
+type Deliverer struct {
+	conf Config
+	c    *samqp.Connection
+}
+
+func New(conf Config, c *samqp.Connection) (*Deliverer, error) {
+	if err := conf.Validate(); err != nil {
+		return nil, err
+	}
+	if c == nil {
+		return nil, fmt.Errorf("ampq connection cannot be nil")
+	}
+
+	return &Deliverer{
+		conf: conf,
+		c:    c,
+	}, nil
+}
+
+func (d *Deliverer) Name() string {
+	return fmt.Sprintf("amqp-%s", d.conf.Exchange.Name)
+}
+
+func (d *Deliverer) Deliver(ctx context.Context, nID uuid.UUID) error {
+	ch, err := d.c.Channel()
+	if err != nil {
+		return &clairerror.ErrDeliveryFailed{err}
+	}
+
+	callback := d.conf.callback
+	callback.Path = path.Join(callback.Path, nID.String())
+
+	cb := notifier.Callback{
+		NotificationID: nID,
+		Callback:       callback,
+	}
+	b, err := json.Marshal(&cb)
+	if err != nil {
+		return &clairerror.ErrDeliveryFailed{err}
+	}
+	msg := samqp.Publishing{
+		ContentType: "application/json",
+		AppId:       "clairV4-notifier",
+		Body:        b,
+	}
+	err = ch.Publish(
+		d.conf.Exchange.Name,
```

Comment:
Is this a queue/topic where message is sent?

Reply:
These are terms used in AMQP spec. https://www.rabbitmq.com/resources/specs/amqp0-9-1.pdf

Reviewing this will clear things up. An exchange routes messages to a queue given some rules.